platform/openide.dialogs/src/org/openide/NotifyDescriptor.java

```diff
@@ -1326,4 +1336,110 @@ public void setSelected(boolean selected) {
             }
         }
     }
+
+    /** Notification providing a composed input of multiple nested selection lists and/or input lines.
+    * @since 7.63
+    */
+    public static final class ComposedInput extends NotifyDescriptor {
+
+        private final List<NotifyDescriptor> inputs = new ArrayList<>();
+        private final Callback callback;
+        private int total;
+
+        /** Construct dialog with the specified title and nested inputs.
+        * @param title title of the dialog
+        * @param total total number of nested inputs
+        * @param callback callback used to create nested inputs
+        * @since 7.63
+        */
+        public ComposedInput(final String title, final int total, final Callback callback) {
+            super(null, title, OK_CANCEL_OPTION, PLAIN_MESSAGE, null, null);
+            this.callback = callback;
+            this.total = total;
+        }
+
+        /**
+         * Total number of nested inputs.
+         * @since 7.63
+         */
+        public int getTotalInputs() {
+            return total;
```

Comment:
The total number of steps is fixed for the whole 'composed' process. But quite often response in a step determines if a subsequent step is relevant at all - often a total number (better: estimated number) of steps changes as user fills inputs.
So I'd allow to change the total steps here, `NotifyDescriptor` has add/removePropertyChangeListener, so the change can be observed + reflected in the UI ?

Reply:
Good point. Fixed.